storage/src/adb/current/ordered.rs

```diff
@@ -0,0 +1,1703 @@
+
+        let last_chunk = self.status.last_chunk();
+        if last_chunk.1 != 0 {
+            hasher.update(last_chunk.0);
+            proof.digests.push(hasher.finalize());
+        }
+
+        Ok((
+            proof,
+            KeyValueProofInfo {
+                key,
+                value,
+                next_key: Some(next_key),
+                loc,
+                chunk,
+            },
+        ))
+    }
+
+    /// Generate and return a proof that the specified `key` does not exist in the db, along with
+    /// the other [KeyValueProofInfo] required to verify the proof. Returns [Error::KeyExists] if
+    /// exclusion can't be proven because the key exists in the db.
+    ///
+    /// # Panic
+    ///
+    /// Panics if there are uncommitted operations.
+    pub async fn exclusion_proof(
+        &self,
+        hasher: &mut H,
+        key: &K,
+    ) -> Result<(Proof<H::Digest>, ExclusionProofInfo<K, V, N>), Error> {
+        assert!(
+            !self.status.is_dirty(),
+            "must process updates before computing proofs"
+        );
+        if self.op_count() == 0 {
+            return Ok((Proof::default(), ExclusionProofInfo::DbEmpty()));
+        }
+        let height = Self::grafting_height();
+        let grafted_mmr = GraftingStorage::<'_, H, _, _>::new(&self.status, &self.any.mmr, height);
+        let (last_chunk, last_bit) = self.status.last_chunk();
+
+        let span = self.any.get_span(key).await?;
+        let (loc, proof_info) = match span {
+            Some((loc, key_data)) => {
+                if key_data.key == *key {
+                    // Cannot prove exclusion of a key that exists in the db.
+                    return Err(Error::KeyExists);
+                }
+                let chunk = *self.status.get_chunk_containing(*loc);
+                (
+                    loc,
+                    ExclusionProofInfo::KeyValue(KeyValueProofInfo {
+                        key: key_data.key,
+                        value: key_data.value,
+                        next_key: Some(key_data.next_key),
+                        loc,
+                        chunk,
+                    }),
+                )
+            }
+            None => {
+                let loc = self
+                    .op_count()
+                    .checked_sub(1)
+                    .expect("db shouldn't be empty");
+                let chunk = *self.status.get_chunk_containing(*loc);
+                (loc, ExclusionProofInfo::Commit((loc, chunk)))
+            }
+        };
+
+        let mut proof = verification::range_proof(&grafted_mmr, loc..loc + 1).await?;
+
+        if last_bit != 0 {
+            hasher.update(last_chunk);
+            proof.digests.push(hasher.finalize());
+        }
+
+        Ok((proof, proof_info))
+    }
+
+    /// Return true if the proof authenticates that `key` currently has value `value` in the db with
+    /// the provided `root`.
+    pub fn verify_key_value_proof(
```

Comment:
This is why I suggest creating a struct for `Proof` and `KeyValueProofInfo` into one type -- it seems that it's impossible to verify the proof without the additional info, which essentially makes it part of the proof.

```suggestion
    /// Return true if the `proof` and `info` authenticate that `key` currently has value `value` in the db with
    /// the provided `root`.
    pub fn verify_key_value_proof(
```

Reply:
I'm not sure I agree but maybe I'm being too pedantic.

I conceptualize the interaction between a verifier and server as:

Verifier: "Give me a proof that K does not exist in the database with root R"
Server: "Here's the stuff you need to prove that K is not in the database with root R"

Personally, I would define all that "stuff" as being the proof, as opposed to just the `Proof` (i.e. MMR size/digests).

I'm fine leaving this as is, though.